src/overlays/actors/ovl_Bg_Po_Event/z_bg_po_event.c

```diff
@@ -21,47 +48,601 @@ const ActorInit Bg_Po_Event_InitVars = {
+                Actor_Kill(&this->dyna.actor);
+                return;
+            }
+            if (thisx->child->child->child == 0) {
+                Actor_Kill(thisx);
+                Actor_Kill(thisx->child);
+                return;
+            }
+            thisx->parent = thisx->child->child->child;
+            thisx->child->child->child->child = thisx;
+        }
+    }
+    thisx->posRot.pos.y = 833.0f;
+    thisx->groundY = func_8003C9A4(&globalCtx->colCtx, &thisx->floorPoly, &dummy, thisx, &thisx->posRot.pos);
+    this->actionFunc = BgPoEvent_BlockWait;
+}
+
+void BgPoEvent_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgPoEvent* this = THIS;
+    ColliderTris* collider;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    this->actorType = (thisx->params >> 8) & 0xF;
+    this->actorIndex = (thisx->params >> 0xC) & 0xF;
+    thisx->params &= 0x3F;
+
+    if (this->actorType >= 2) {
+        collider = &this->collider1;
+        Collider_InitTris(globalCtx, collider);
+        Collider_SetTris(globalCtx, collider, &this->dyna.actor, &sTrisInit, this->colliderItems);
+        if (Flags_GetSwitch(globalCtx, thisx->params)) {
+            Actor_Kill(&this->dyna.actor);
+        } else {
+            BgPoEvent_InitPaintings(this, globalCtx);
+        }
+    } else {
+        DynaPolyInfo_SetActorMove(&this->dyna, 0);
+        if (Flags_GetSwitch(globalCtx, thisx->params)) {
+            Actor_Kill(&this->dyna.actor);
+        } else {
+            BgPoEvent_InitBlocks(this, globalCtx);
+        }
+    }
+}
+
+void BgPoEvent_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgPoEvent* this = THIS;
+
+    if (this->actorType >= 2) {
+        Collider_DestroyTris(globalCtx, &this->collider1);
+    } else {
+        DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+        if ((this->actorType == 1) && (gSaveContext.timer1Value > 0)) {
+            gSaveContext.timer1State = 0xA;
+        }
+    }
+}
+
+void BgPoEvent_BlockWait(BgPoEvent* this, GlobalContext* globalCtx) {
+
+    this->dyna.actor.posRot.pos.y = 833.0f;
+    if (puzzleState == 0x3F) {
+        if (this->actorType == 1) {
+            func_800800F8(globalCtx, 0xC4E, 0x41, NULL, 0);
+        }
+        this->timer = 45;
+        this->actionFunc = BgPoEvent_BlockShake;
+    } else if (this->dyna.actor.xzDistFromLink > 50.0f) {
+        if (this->actorType != 1) {
+            puzzleState |= (1 << this->actorIndex);
+        } else {
+            puzzleState |= 0x10;
+        }
+    } else if (this->actorType != 1) {
+        puzzleState &= ~(1 << this->actorIndex);
+    } else {
+        puzzleState &= ~0x10;
+    }
+}
+
+void BgPoEvent_BlockShake(BgPoEvent* this, GlobalContext* globalCtx) {
+
+    DECR(this->timer);
+    if (this->timer < 15) {
+        this->dyna.actor.posRot.pos.x = this->dyna.actor.initPosRot.pos.x + 2.0f * ((this->timer % 3) - 1);
+        if (!(this->timer % 4)) {
```

Comment:
I think ` == 0` makes more sense when checking actual numerical values

Reply:
You'd think that, but it actually doesn't match. For modulo operations with a power of 4, setting it equal to zero and using logical not aren't the same thing. In the former, the compiler knows it doesn't have to care about the sign of the result, but it doesn't know that with logical not.